packages/context/src/context.ts

```diff
@@ -64,65 +112,167 @@ export class Context {
   }
 
+  /**
+   * Iterate all contexts by breadth first traversal of the graph following
+   * `parents`. For example, with the following context graph:
+   * - reqCtx -> [serverCtx, connectorCtx]
+   * - serverCtx -> [appCtx]
+   * - connectorCtx -> [appCtx]
+   * `reqCtx.contexts()` returns an iterator of `[reqCtx, serverCtx,
+   * connectorCtx, appCtx]`.
+   */
+  protected *contexts(): IterableIterator<Context> {
+    const visited: Set<Context> = new Set();
+    const queue: Context[] = [];
+    // Enqueue the current context
+    queue.push(this);
+    while (queue.length) {
+      // Dequeue the head context
+      const c = queue.shift()!;
+      // Check if the context has been added
+      if (!visited.has(c)) {
+        visited.add(c);
+        yield c;
+        // Enqueue the parent contexts
+        queue.push(...c.parents);
+      }
+    }
+  }
+
+  /**
+   * Visit all contexts in the graph. It first calls the `visitor` function to
+   * produce a `result` against the current context. If the `result.done` flag
+   * is set to `true`, the method returns `result.value`. Otherwise, it iterates
+   * through all parent contexts and uses the `reducer` function to reduce the
+   * value from the parent binding chain against the last value. To avoid
+   * circular or duplicate visits, a set is used to keep track of all context
+   * objects that have been visited.
+   *
+   * @param visitor A function to produce a result from the context object
+   * locally without consulting the parents.
+   * @param reducer An optional function to reduce the value from the value
+   * from a parent binding againt the last result. If not present, a default
+   * reducer is used to use the current value to override the last one.
+   * @param visited An optional set to keep track of visited contexts.
+   */
+  protected visitAllContexts<T>(
+    visitor: (ctx: Context) => {done: boolean; value: T},
+    reducer?: (previousValue: T, currentValue: T) => T,
+  ): T {
+    const contexts = this.contexts();
+    // The default reducer keeps the current value
+    const reduceFn = reducer || ((prev, current) => current);
+    let prevVal: T | undefined = undefined;
+    for (const ctx of contexts) {
+      const result = visitor(ctx);
+      prevVal =
+        prevVal === undefined ? result.value : reduceFn(prevVal, result.value);
+      if (result.done) break;
+    }
+    // At least one instance of context exists in the loop above. The `prevVal`
+    // should have been set to an instance of type T at this point.
+    return prevVal!;
+  }
+
   /**
    * Check if a key is bound in the context or its ancestors
    * @param key Binding key
    */
   isBound(key: string): boolean {
-    if (this.contains(key)) return true;
-    if (this._parent) {
-      return this._parent.isBound(key);
-    }
-    return false;
+    const visitor = (ctx: Context) => {
+      const found = ctx.contains(key);
+      return {done: found, value: found};
```

Comment:
I have bad feeling about `return {done, value}` style, it feels like we are manually reimplementing Iterable contract. Can we find a way how to leverage generator functions for this?
 
 - `return {done: false, value: found}` should be replaced with `yield found`
 -  `return {done: true}` should be replaced with `return;`

My experience with generator functions was limited so far, please let me know if my expectations are not matching the reality of ES6.